apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/DLQGrpcRdaSource.java

```diff
@@ -0,0 +1,227 @@
+    }
+
+    return totalProcessed;
+  }
+
+  @VisibleForTesting
+  Processor dlqProcessingLogic(
+      RdaSink<TMessage, TClaim> sink, MessageError.ClaimType type, Set<Long> sequenceNumbers) {
+    return () -> {
+      ProcessResult processResult = new ProcessResult();
+      int processed = 0;
+      final String apiVersion = caller.callVersionService(channel, callOptionsFactory.get());
+
+      for (final long startingSequenceNumber : sequenceNumbers) {
+        log.info(
+            "calling API for {} claims starting at sequence number {}",
+            claimType,
+            startingSequenceNumber);
+
+        final GrpcResponseStream<TMessage> responseStream =
+            caller.callService(channel, callOptionsFactory.get(), startingSequenceNumber);
+        final Map<Object, TMessage> batch = new LinkedHashMap<>();
+
+        try {
+          if (responseStream.hasNext()) {
+            setUptimeToReceiving();
+            final TMessage result = responseStream.next();
+            metrics.getObjectsReceived().mark();
+
+            if (sequencePredicate.test(startingSequenceNumber, result)) {
+              batch.put(sink.getDedupKeyForMessage(result), result);
+              processed += submitBatchToSink(apiVersion, sink, batch);
+
+              if (processed > 0 && dao.delete(startingSequenceNumber, type) > 0) {
+                log.info(
+                    "{} claim with sequence ({}) processed successfully, removed DLQ entry",
+                    claimType,
+                    startingSequenceNumber);
+              }
+            }
+
+            responseStream.cancelStream("No further messages need to be ingested.");
+          }
+        } catch (GrpcResponseStream.StreamInterruptedException ex) {
+          // If our thread is interrupted we cancel the stream so the server knows we're done
+          // and then shut down normally.
+          responseStream.cancelStream("shutting down due to InterruptedException");
+          processResult.setInterrupted(true);
+        } catch (Exception e) {
+          // If we failed to process the claim, it stays in the DLQ, nothing to do.
+          log.error(
+              "Failed to process "
+                  + claimType
+                  + " message with sequence: "
+                  + startingSequenceNumber,
+              e);
+        }
+
+        sink.shutdown(Duration.ofMinutes(5));
+        processed += sink.getProcessedCount();
+      }
+
+      processResult.setCount(processed);
+      return processResult;
+    };
+  }
+
+  /** Utility Data Access Object (DAO) class for querying the database. */
+  @VisibleForTesting
+  @RequiredArgsConstructor
+  static class DLQDao {
+
+    private final EntityManager entityManager;
+
+    public List<MessageError> findAllMessageErrors() {
+      return entityManager
+          .createQuery("select error from MessageError error", MessageError.class)
+          .getResultList();
```

Comment:
Why not add a where clause for the `claim_type` here instead of filtering in the calling method?